### DISCUSSION TITLE
First Query after Flow Launcher starts up is very slow

### DISCUSSION THREAD

**Discussion by kubalav** [Flow-Launcher/Flow.Launcher#586] [Q&A]
I think my fist query take much time (cca 8 sec). After that is blazing fast.
I type "git" and it freeze at G.

https://user-images.githubusercontent.com/37414585/126500407-8de15b8a-470b-473c-a8e4-c32bcdd74d7a.mp4

Can it caused by this error in log?
`Possible interpretation of the error: Unable to get logo for Microsoft.MicrosoftEdge_8wekyb3d8bbwe!MicrosoftEdge from  and located in C:\Windows\SystemApps\Microsoft.MicrosoftEdge_8wekyb3d8bbwe`

**Comment by taooceros** [comment]
From the video, the slowness shouldn't come from cost of querying, because we can see the suggested title. Probably is because some wpf window initialization or the issue discussed in #553.

The slowness shouldn't come from the error you show, because it is supposed to use a new thread to retrieve image when not exist. However, it is possible that the image is from cache but hasn't got loaded fully, which may lead to some weird issue possibly.

Does it happens before 1.8.0 or upgrading to 1.8.0 cause the issue? By the way, does this happen everytime restarting Flow?

**Reply by kubalav** [reply]
It happens after upgrading to 1.8.0. This is repeated every time I start Flow launcher.

**Reply by pc223** [reply]
@taooceros Seems related to #554, I think we should try to push this PR to 1.8.1

**Reply by pc223** [reply]
@kubalav Can you try the build of #554  here to see if the delay go away?

https://ci.appveyor.com/project/JohnTheGr8/flow-launcher/builds/39955889/artifacts

**Reply by pc223** [reply]
@taooceros I confirm with the 1.8.0 official release, the delay I describe in #553 still exists on my machine

**Reply by kubalav** [reply]
> @kubalav Can you try the build of #554 here to see if the delay go away?
> 
> [ci.appveyor.com/project/JohnTheGr8/flow-launcher/builds/39955889/artifacts](https://ci.appveyor.com/project/JohnTheGr8/flow-launcher/builds/39955889/artifacts)

I tried #554 build and delay persist.

**Reply by pc223** [reply]
> > @kubalav Can you try the build of #554 here to see if the delay go away?
> > [ci.appveyor.com/project/JohnTheGr8/flow-launcher/builds/39955889/artifacts](https://ci.appveyor.com/project/JohnTheGr8/flow-launcher/builds/39955889/artifacts)
> 
> I tried #554 build and delay persist.

This is strange, #554 fix the delay on my machine, @taooceros any idea how to find out why?

**Reply by taooceros** [reply]
@kubalav what about waiting a while and try querying? Does the result appears as the same?

**Reply by kubalav** [reply]
I tried build #554 again, but with a clean install. Everything is working as expected 😀. I tried to copy UserData to this build. Result? Freezing. Maybe some misconfiguration. So, I have to configured it again, but most important, it's working as it should.

Thank you, guys!

@taooceros @pc223

**Reply by taooceros** [reply]
Nice to hear that. Could you share the configure to see what has gone wrong?

**Reply by kubalav** [reply]
@taooceros
Do you mean Settings.json?

**Reply by pc223** [reply]
> @taooceros
> Do you mean Settings.json?

Please provide as much as possible info, logs file, Settings.json, list of plugins and version of them, your system spec...

With your report and discussion in #553, we know that #554 is only a temporary solution, so it's bound to happen again with other users, so we need your info to try to reproduce the problem 😄

**Reply by kubalav** [reply]
Procesor i5-5200U CPU @ 2.20GHz 4 cores
RAM	8,00 GB
Win 10 21H1

Log file:
[2021-07-24.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/6874185/2021-07-24.txt)
Settings.json
[Settings.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/6874192/Settings.txt)

**Reply by pc223** [reply]
> Procesor i5-5200U CPU @ 2.20GHz 4 cores
> RAM 8,00 GB
> Win 10 21H1
> 
> Log file:
> [2021-07-24.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/6874185/2021-07-24.txt)
> Settings.json
> [Settings.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/6874192/Settings.txt)

The things I notice is you're using Everything 1.4.9, latest is 1.5.1 and you use a lot of * (asterisk) action keywords, meaning the results count should be pretty high. Probably the delay related to those things. I'll do further tests to investigate later.

**Reply by jjw24** [reply]
Also got a freeze with nothing logged for #554, i was not using any external plugins

**Reply by taooceros** [reply]
Actually the everything plugin won't create a lot of result because the result count is 100 for both 1.4.9 and 1.5.1. Probably the high count comes from program plugin.

**Comment by pc223** [comment]
Okay, continue the investigation :smile:

I got a hunch, maybe it got something to do with the Program plugin, maybe a specific software on my machine (and @kubalav) that causing the delay when loading the icon of that program or something.

@kubalav seems can reproduce the delay when upgrade to `1.8.0`? What was your Flow version before you upgraded to `1.8.0`? Do you experience the delay with that old version right now?

@jjw24 seems not able to reproduce the delay

@taooceros Do you able to reproduce the delay on dev branch?

@Flow-Launcher/team Anyone else can reproduce this delay on dev branch or latest releases?

Steps to reproduce:

Suppose you setup action keywords like this, `pr <query>` for Program plugin and `cp <query>` for ControlPanel plugin.

1. Kill Flow (right click taskbar icon to exit) and launch it back. (Have to kill Flow every time to see the delay)
2. Use `pr a` to search for a program, you should see a delay (around 3-5 secs)
3. Kill Flow and launch it back
4. Use `cp a` to search for a control panel item, the results should appear immediately, continue, use `pr a` to search for a program, the results should appear immediately. (No delay, like a "block" being cleared by `cp a`)

**Reply by jjw24** [reply]
i do keep a very minimal set of apps on my laptop though haha so maybe.

the steps to repro, are they the same for debug and release?

**Comment by kubalav** [comment]
I thought the problem occurred after the upgrade from 1.7.2 to version 1.8.0
I can reproduce problem with old version too. If I set action keyword for Program plugin, then search query work flawlessly (maybe 1 sec delay).

**Reply by pc223** [reply]
Ok, so with the old version `1.7.2` you still experience the same behavior as `1.8.0`?

And if you set a action keyword for Program plugin, then it doesn't delay? Either with `1.8.0` or the old one?

**Reply by kubalav** [reply]
Version 1.7.2 have the same delay than 1.8.0. If action keyword is set for Plugin program it doesn't delay with 1.7.2 and 1.8.0.

**Reply by pc223** [reply]
> Version 1.7.2 have the same delay than 1.8.0. If action keyword is set for Plugin program it doesn't delay with 1.7.2 and 1.8.0.

That is weird, so your case is not exactly same as my.

Can you try to isolate the plugin causing problem, like set action keyword for all of the plugins, and kill Flow and query a plugin and kill Flow and query a plugin...

**Reply by deefrawley** [reply]
Don't the logs show the startup time for each plugin? Wouldn't that help? e.g.
```08:06:30.5064+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <Plugin Indicator> is <39ms>
08:06:30.5064+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <Process Killer> is <49ms>
08:06:30.5573+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <Explorer> is <130ms>
08:06:30.5745+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <System Commands> is <49ms>```

**Reply by pc223** [reply]
> Don't the logs show the startup time for each plugin? Wouldn't that help? e.g.
> 
> ```
> 08:06:30.5064+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <Process Killer> is <49ms>
> 08:06:30.5573+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <Explorer> is <130ms>
> 08:06:30.5745+10:00 - INFO  - PluginManager.InitializePlugins - Total init cost for <System Commands> is <49ms>```
> ```

The problem seems to related to this line, not the startup time:

https://github.com/Flow-Launcher/Flow.Launcher/blob/4219dce3799df2f30180a66ec579e42aedd71314/Flow.Launcher/ViewModel/ResultsViewModel.cs#L261

I don't think of any way to log out the timing after that line.

**Reply by jjw24** [reply]
what about disabling Programs plugin, as well as trying the test disabling all plugins and enabling only programs plugin. finally disable some of the programs via the programs list.

**Comment by onesounds** [comment]
Is this problem solved?
There is a very slow response only in the first query. From the second query, it works comfortably.
It can only be reproduced by turning off and on the program.
This problem is difficult to realize when you get used to the flow. I can definitely feel the difference when compared to wox or powertoy.

**Reply by onesounds** [reply]
**Similar Problems**
https://github.com/microsoft/PowerToys/issues/7208
https://github.com/microsoft/PowerToys/pull/11364
https://github.com/microsoft/PowerToys/pull/5748
https://github.com/microsoft/PowerToys/issues/4918
https://github.com/microsoft/PowerToys/issues/12968
https://github.com/microsoft/PowerToys/issues/3830

**Comment by kubalav** [comment]
The issue is still persist.

**Comment by spookyuser** [comment]
I'm still experiencing this :/

**Comment by aaronedev** [comment]
yea same issue here
